alert_generator/specification.md

```diff
@@ -0,0 +1,398 @@
+  "endsAt": "<RFC3339Millis time>",
+  "generatorURL": "<string>"
+}
+```
+
+* `labels`: MUST be present. The labels uniquely identify an alert.
+* `annotations`: MUST be present IF the alert has annotations. Annotations provide additional details about the alert which can change over time for the same alert.
+* `startsAt`: SHOULD be present. It is the time when the alert was triggered.
+* `endsAt`: SHOULD be present. It is the time when the alert MUST be considered `inactive`. Note that future alert updates MAY change this value.
+* `generatorURL`: MAY be present. It is a URL that takes the user to the query page for the source expression of the alert.
+
+## Alerting Rules
+
+### The Format
+
+The alert-generator MUST accept the Prometheus style alerting rules configuration as described in the [v2.30 docs for Alerting Rules](https://prometheus.io/docs/prometheus/2.30/configuration/alerting_rules/) with the following structure. It MUST be in either YAML format or an equivalent JSON format. Alert-generator MAY accept them as files on disk or via an API.
+
+```yaml
+groups:
+  [ - <rule_group> ]
+```
+
+**`<rule_group>`**
+```yaml
+# The name of the group. MUST be unique within a file.
+name: <string>
+
+# How often rules in the group are evaluated.
+[ interval: <duration> | default = 1m ]
+
+rules:
+  [ - <rule> ... ]
+```
+
+**`<rule>`**
+```yaml
+# The name of the alert. MUST be a valid label value.
+alert: <string>
+
+# The PromQL expression to evaluate. Every evaluation cycle this is
+# evaluated at the current time, and all resultant time series become
+# pending/firing alerts.
+expr: <string>
+
+# Alerts are considered firing once they have been returned for this long.
+# Alerts which have not yet fired for long enough are considered pending.
+[ for: <duration> | default = 0s ]
+
+# Labels to add or overwrite for each alert.
+labels:
+  [ <labelname>: <template_string> ]
+
+# Annotations to add to each alert.
+annotations:
+  [ <labelname>: <template_string> ]
+```
+
+Example config
+```yaml
+groups:
+- name: example
+  rules:
+  - alert: HighRequestLatency
+    expr: job:request_latency_seconds:mean5m{job="myjob"} > 0.5
+    for: 10m
+    labels:
+      severity: page
+    annotations:
+      summary: High request latency
+  - alert: VeryHighRequestLatency
+    expr: job:request_latency_seconds:mean5m{job="myjob"} > 1
+```
+### Constraints
+1. 
+2. Results of a rule evaluation MUST be available for any subsequent rules in the group during the same evaluation cycle and the order of rules MUST be the same as provided in the config.
+3. Labels and annotations in alerting rules MUST support all template variables and functions as described in the [v2.30 template reference](https://prometheus.io/docs/prometheus/2.30/configuration/template_reference/) for the values except `graphLink` and `tableLink`. `graphLink` and `tableLink` MAY be supported if the sample querier supports having a UI link for graph and table respectively.
```

Comment:
> what happens if it's not supported

I believe you are talking about `graphLink` and `tableLink`. I think empty string is a good idea, I will mention or. Or do you think it should be something else?

Reply:
I'm talking about graphLink&tableLink.

I think the main thing is that the config doesn't get rejected, otherwise the user will have portability issues.

Reply:
Right. With the last commit it should be clear that the config must not be rejected and should result in an empty string if present.